packages/react-router/lib/types/route-module.ts

```diff
@@ -260,3 +260,13 @@ type __tests = [
     >
   >
 ];
+
+/**
+ * A brand that can be applied to a type to indicate that it will serialize
+ * to a specific type when transported to the client from a loader.
+ * Only use this if you have additional serialization/deserialization logic
+ * in your application.
+ */
+export type SerializesTo<T> = {
+  $__RR_SerializesTo?: [T];
```

Comment:
@phryneas I was expecting `$__RR_SerializesTo?: T` without the wrapping `[]`. Does wrapping in `[]` come with any benefits over unwrapped `T`?

Reply:
It's probably not a big deal, but It's slightly more safe around `undefined`:

```js
{
  interface WithoutWrapping<T> {
    foo?: T;
  }

  type SerializesTo<T> = T extends WithoutWrapping<infer F> ? F : "other";

  type Test1 = SerializesTo<WithoutWrapping<number>>;
  //    ^?  type Test1 = number

  type Test2 = SerializesTo<WithoutWrapping<number|undefined>>;
  //    ^?  type Test2 = number | undefined

  // this could happen if for some reason the TS compiler inlines the `WithoutWrapping` type
  // when creating declarations for a third-party-package
  // an explicitly stated `undefined` would get lost
  type Test3 = SerializesTo<{ foo?: number|undefined }>;
  //    ^?  type Test3 = number
}
```

Reply:
Could you explain why the brand is optional in the first place?

Why not have:

```ts
export type SerializesTo<T> = {
  __ReactRouter_SerializesTo: T;
}
```

Reply:
For assignability reasons - if it were non-optional you'd need an `as any` or `as unknown` when using it:

```ts
type SerializesTo<T> = {
  __ReactRouter_SerializesTo?: [T];
};


const foo: number & SerializesTo<string> = 5

type SerializesToNonOptional<T> = {
  __ReactRouter_SerializesTo: [T];
};

// this errors
const bar: number & SerializesToNonOptional<string> = 5
```

Reply:
Also, just thinking of this - this is not actually there at runtime, so it keeps the type kinda "correct" in line with what you'd expect at runtime.